Fix GH-9296: incorrect `ksort(..., SORT_REGULAR)` behaviour on arrays with numeric and string keys

DrDet opened the pull request:
The comparator function used at `ksort` or `krsort` in `SORT_REGULAR` mode need to be consistent with basic comparison rules. 
These rules were changed in PHP-8.0 for numeric strings, but comparator used at `ksort` mistakenly kept the old behaviour. 
It leads to inconsistent situations, when after `ksort` the first key is **GREATER** than some of the next ones by according to the basic comparison operators.

Simple example:
```php
<?php

$array = ["600" => 0, "aaa" => 1];
ksort($array, SORT_REGULAR);
var_dump($array);
var_dump(array_key_first($array) <=> array_key_last($array));
```
In php7.4 we get:
```
array(2) {
  ["aaa"]=>
  int(1)
  [600]=>
  int(0)
}
int(-1)
```
That's OK, because it's consistent with php7.4 comparison rules.

In php8.0 the output is:
```
array(2) {
  ["aaa"]=>
  int(1)
  [600]=>
  int(0)
}
int(1)
```
Which is not OK, because it is not consistent with new php8.0 comparison rules.

After this fix the output is:
```
array(2) {
  [600]=>
  int(0)
  ["aaa"]=>
  int(1)
}
int(-1)
```
Which is consistent with php8.0 comparison rules.

See #9296

cmb69 commented on ext/standard/tests/array/krsort_variation9.phpt:
```diff
@@ -82,22 +82,22 @@ array(5) {
 bool(true)
 array(3) {
+  ["c"]=>
+  string(5) "apple"
   ["a"]=>
   string(6) "orange"
   [0]=>
   string(6) "banana"
-  ["c"]=>
-  string(5) "apple"
 }
```
This looks wrong; the old result was correct. Same for the other `krsort()` test below.

DrDet replied:
The order of keys by according to the old result: `["a", 0, "c"]`
Let's check it with a spaceship operator:
`print_r(["a" <=> 0, 0 <=> "c"]);` gives the following output:
```
Array
(
    [0] => 1
    [1] => -1
)
```
Which by definition means that it's not sorted and therefore the old result is incorrect.

Let's do the same with the new result: `["c", "a", 0]`
`print_r(["c" <=> "a", "a" <=> 0]);` gives us the following:
```
Array
(
    [0] => 1
    [1] => 1
)
```
Which means that it's sorted and actually correct.

cmb69 replied:
Of course, you're right! I was confused; sorry for the noise.

cmb69 reviewed (commented):
Ugh! I'm pretty confused that this apparently has not been reported before.

dstogov commented:
I think this is a bug and it's better to report it at https://github.com/php/php-src/issues

I would prefer, to keep the old (optimized) comparison code at least for numeric+numeric and string+string keys, and fallback to ``zend_compare()`` only for rare numeric+string and string+numeric cases.

It would be also great to have a separate test case with a reference to the issue report.

DrDet commented:
> I think this is a bug and it's better to report it at https://github.com/php/php-src/issues

Done with the bug report, see #9296 

> I would prefer, to keep the old (optimized) comparison code at least for numeric+numeric and string+string keys, and fallback to `zend_compare()` only for rare numeric+string and string+numeric cases.
> 

Fixed at [aa125c8](https://github.com/php/php-src/pull/9293/commits/aa125c8f848920dbf11cac10ca69f44ec2f65773)

> It would be also great to have a separate test case with a reference to the issue report.

Added at [23fc6d4](https://github.com/php/php-src/pull/9293/commits/23fc6d43b9cda1296776d0687988b09c6c0faa30)

dstogov reviewed (approved):
Now this looks fine.

cmb69 commented:
Thank you!
